streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -824,27 +834,42 @@ long runOnce(final long recordsProcessedBeforeCommit) {
         }
 
         if (records != null && !records.isEmpty()) {
-            streamsMetrics.pollTimeSensor.record(computeLatency(), timerStartedMs);
+            streamsMetrics.pollTimeSensor.record(computeLatency(), now);
             addRecordsToTasks(records);
         }
 
         if (taskManager.hasActiveRunningTasks()) {
-            final long totalProcessed = processAndMaybeCommit(recordsProcessedBeforeCommit);
-            if (totalProcessed > 0) {
-                final long processLatency = computeLatency();
-                streamsMetrics.processTimeSensor.record(processLatency / (double) totalProcessed, timerStartedMs);
-                processedBeforeCommit = adjustRecordsProcessedBeforeCommit(
-                        recordsProcessedBeforeCommit,
-                        totalProcessed,
-                        processLatency,
-                        commitTimeMs);
-            }
+            taskManager.maybeEnforceProcess(now);
+
+            /*
+             * Within an iteration, after N (N initialized as 1 upon start up) round of processing one-record-each on the applicable tasks, check the current time:
+             *  1. If it is time to commit, do it;
+             *  2. If it is time to punctuate, do it;
+             *  3. If elapsed time is close to consumer's max.poll.interval.ms, end the current iteration immediately.
+             *  4. If none of the the above happens, increment N.
+             *  5. If one of the above happens, half the value of N.
+             */
+            long totalProcessed;
+            long timeSinceLastPoll;
+
+            do {
+                totalProcessed = processAndMaybeCommit();
+                timeSinceLastPoll = Math.max(now - lastPollMs, 0);
+
+                if (maybePunctuate() || maybeCommit()) {
+                    numIterations = numIterations > 1 ? numIterations / 2 : numIterations;
+                } else {
+                    numIterations++;
+                }
+            } while (totalProcessed > 0 && timeSinceLastPoll < maxPollTimeMs / 2);
         }
 
-        punctuate();
-        maybeCommit(timerStartedMs);
-        maybeUpdateStandbyTasks(timerStartedMs);
-        return processedBeforeCommit;
+        // even if there is not data to process in this iteration, still need to check if commit / punctuate is needed
```

Comment:
Good catch.

The deliberation was that even though "there is no data processed", not "there is no active tasks" as the original check is `State == RUNNING` :) Note that although `taskManager.hasActiveRunningTasks()` returns true, we may still not process any data (i.e. `totalProcessed` == 0 and we break the loop immediately).

But with the new condition, we will always execute `if (maybePunctuate() || maybeCommit())` anyways, so we only need to do `maybeUpdate` followed by a `maybeCommit` again.

Reply:
Actually thinking about this, I feel it is better to separate the standby tasks from active tasks in maybeCommit as otherwise we are doomed to waste some CPUs doing either one of them. Will refactor the code a bit more.